crates/astria-conductor/src/block_verifier.rs

```diff
@@ -0,0 +1,583 @@
+                res_address
+            );
+        }
+
+        Ok(())
+    }
+
+    /// performs various validation checks on the SequencerBlock received from either gossip or
+    /// Celestia.
+    ///
+    /// checks performed:
+    /// - the proposer of the sequencer block matches the expected proposer for the block height
+    ///   from tendermint
+    /// - the signer of the SignedNamespaceData the proposer
+    /// - the signature is valid
+    /// - the root of the markle tree of all the header fields matches the block's block_hash
+    /// - the root of the merkle tree of all transactions in the block matches the block's data_hash
+    /// - validate the block was actually finalized; ie >2/3 stake signed off on it
+    pub(crate) async fn validate_sequencer_block(&self, block: &SequencerBlock) -> Result<()> {
+        // sequencer block's height
+        let height = block
+            .header
+            .height
+            .parse::<u64>()
+            .wrap_err("failed to parse height")?;
+
+        // get validator set for this height
+        let validator_set = self
+            .tendermint_client
+            .get_validator_set(height - 1)
+            .await
+            .wrap_err("failed to get validator set")?;
+
+        // find proposer address for this height
+        let expected_proposer_address = validator_set
+            .get_proposer()
+            .wrap_err("failed to get proposer from validator set")?
+            .address;
+
+        // check if the proposer address matches the sequencer block's proposer
+        let received_proposer_address = bech32::encode(
+            "metrovalcons",
+            block.header.proposer_address.0.to_base32(),
+            Variant::Bech32,
+        )
+        .wrap_err("failed converting bytes to bech32 address")?;
+
+        if received_proposer_address != expected_proposer_address {
+            bail!(
+                "proposer address mismatch: expected {}, got {}",
+                expected_proposer_address,
+                received_proposer_address
+            );
+        }
+
+        // validate that commit signatures hash to header.last_commit_hash
+        match calculate_last_commit_hash(&block.last_commit) {
+            Hash::Sha256(calculated_last_commit_hash) => {
+                let Some(last_commit_hash) = block.header.last_commit_hash.as_ref() else {
+                    bail!("last commit hash should not be empty");
+                };
+
+                if calculated_last_commit_hash.as_slice() != last_commit_hash.0 {
+                    bail!("last commit hash in header does not match calculated last commit hash");
+                }
+
+                // verify that the validator votes on the previous block have >2/3 voting power
+                let last_commit = block.last_commit.clone();
+                let chain_id = block.header.chain_id.clone();
+                tokio::task::spawn_blocking(move || -> Result<()> {
```

Comment:
I'm not sure what is happening here. Why are spawning a blocking task and then awaiting it immediately?

Reply:
@SuperFluffy ?